### PR TITLE
word2vec Genderify

### PR DESCRIPTION
**Pull request by RoyRin**
* Transformation to add masculinity or femininity to each of the nouns in a sentence.
* 1. Take each nouns in a sentence, compute their word2vec vectors
* 2. add `man`-word2vec-vector to them, (and `woman`-vector) 
* 3. then to replace the noun in the sentence with the closest word-vector 
   to (`word-vector` + `man/woman`-vector ).

### CODE REVIEW

File: transformations/word2vec_genderify/test.json [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    "test_cases": [
+        {
+            "class": "GenderifyOperation",
+            "inputs": {
+                "sentence": "Andrew finally returned the French book to Chris that I bought last week"
+            },
+            "outputs": [
+                {
+                    "sentence": "andrew finally returned the french memoir to chris that i bought last week"
+                }
```
[the remark this answers is not in the document]
Comment (RoyRin) [reply transformations/word2vec_genderify/test.json]: done.

**Comment by marco-digio** [comment]
Hi @RoyRin 
I was assigned as one of your reviewers. 
Thanks for this contribution! 

I am not sure to understand why this transformation could be helpful. Do you have any reference about works related to this? I have usually seen transformations that, given a male noun, adding `woman` vector and subtracting `man` vector, results in the feminine version of the original noun, but I have never seen intuitive explainations of adding `man` or `woman` vector to a noun. 
Moreover, the code prioritize manify_function to womanify_function, see the comment below. 

Correctness: All checks have failed
Interface: I think you have chosen the correct interface(s).
Applicable Tasks & Keywords: Why not TEXT_CLASSIFICATION? Keywords are good. 
Specificity: The change is general
Novelty: This transformation is not yet implemented in NL-Augmenter.
Adding New Libraries: inflect==5.3.0, gensim==4.0.1, spacy~=3.0.0, toolz==0.11.1
Description: The README is clear and explains what the transformation is aiming to do.
Data and code source: Missing references to related works. 
Test Cases: 4 test cases have been added.
Evaluating Robustness: No robustness evaluation has been performed.
Languages other than English:   No

File: transformations/word2vec_genderify/transformation.py [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+        swapped_sentence = swap_out_words(sentence, str(noun),
+                                          womanify_function, inflect_engine)
+        if sentence != swapped_sentence:  # only add if there are any changes
+            swapped_sentences.append(swapped_sentence)
+            count += 1
+        if count > max_outputs:
+            break
+    if count == 0:
+        return [sentence]
+    return swapped_sentences[:max_outputs]
```
Comment (marco-digio) [inline transformations/word2vec_genderify/transformation.py:205]: Here `manify_function` is prioritized to `womanify_function`.
For example, if `max_outputs ` = 1, and `swapped_sentence ` from `manify_function` is different from the original sentence, the code always outputs that, without chance of selecting a possible  `swapped_sentence ` from `womanify_function `. I suggest to randomize this.

**Review by ashish3586 — APPROVED** [review]

File: transformations/word2vec_genderify/README.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+This transformation could be used to improve datasets for models that seek to improve their implicit bias (i.e. if the dataset comes from a source that always uses the masculine or the feminine version of nouns). 
+
+Further, this same model can be used to add other kinds of words (beyond "man" and "woman") to each of the nouns. Which may allow for a more balanced dataset with regards to equitable vocab expression - i.e. add "young" and "old" vectors to all nouns or "fast" and "slow".
+
+## What are the limitations of this transformation?
+
+The transformation relies on the accuracy of word2vec, and that word2vec vector addition preserves analogy well. This process is hard to predict and edge-cases are difficult to account for with hard-and-fast rules. This transformation does not check whether the transformed sentence makes logical sense; while most cases it works, it's difficult to ensure.
+
+A foreign name like "Dev" which is someone's name, may be interpreted as a developer, and changed like a noun.
+
```
Comment (ashish3586) [inline transformations/word2vec_genderify/README.md:39]: Please add Robustness evaluation results. You can refer to PR #218.

File: transformations/word2vec_genderify/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,5 @@
+
+inflect==5.3.0
+gensim==4.0.1
+spacy~=3.0.0
```
Comment (AbinayaM02) [inline transformations/word2vec_genderify/requirements.txt:4]: Spacy can be removed, since it's present in the main requirements file.

**Comment by AbinayaM02** [comment]
Hi @RoyRin: Please add your transformation name in the test/mapper.py file in the appropriate dictionary. I think the build is failing because of old code. Please pull the code and let's check if the code passes through.

**Comment by mille-s** [comment]
Interesting tranformation, @RoyRin  please address the comments above!

**Comment by kaustubhdhole** [comment]
Feel free to reopen after fixing the suggestions!

**Pull request closed without merging**